File: Code_cpp/101_tree-easy_对称二叉树.cpp
/*
 * @lc app=leetcode.cn id=101 lang=cpp
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// struct TreeNode {
//     int val;
//      TreeNode *left;
//      TreeNode *right;
//      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
// };

/* 解法1:递归DFS */
/* class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(root==nullptr)
            return true;
        return (isMirror(root->left,root->right));
        
    }
    bool isMirror(TreeNode* t1, TreeNode*t2){
        if(t1==nullptr && t2==nullptr)
            return true;
        if(t1==nullptr || t2==nullptr)
            return false;
        return (t1->val==t2->val
            && isMirror(t1->left,t2->right)
            && isMirror(t1->right,t2->left));
    }
}; */

/* 解法2:迭代BFS */
// #include <queue>
// using namespace std;
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        q.push(root);

        while (!q.empty())
        {
            TreeNode* t1 = q.front();q.pop();
            TreeNode* t2 = q.front();q.pop();
            if(t1 == nullptr && t2 == nullptr)
                continue;
            if(t1 == nullptr || t2 == nullptr
                || t1->val != t2->val)
                return false;
            q.push(t1->right);
            q.push(t2->left);
            q.push(t2->right);
            q.push(t1->left);
        }
        return true;        
    }
};
// @lc code=end

File: Code_java/101_tree-easy_对称二叉树.java
import java.util.LinkedList;
import java.util.Queue;

import javax.management.Query;

import sun.reflect.generics.tree.Tree;

/*
 * @lc app=leetcode.cn id=101 lang=java
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

/**
 * 递归法
 */
// class Solution {
// /**
// * helper: Is two trees mirror?
// *
// * @param left
// * @param right
// * @return
// */
// private boolean helper(TreeNode left, TreeNode right) {
// if(left == null || right==null){
// if(left==null&&right==null){
// return true;
// }
// return false;
// }
// return (left.val == right.val
// && helper(left.left, right.right)
// && helper(left.right, right.left));
// }

// public boolean isSymmetric(TreeNode root) {
// if(root==null){
// return true;
// }
// return helper(root.left, root.right);
// }
// }

/**
 * 迭代法
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode first = queue.remove();
            TreeNode second = queue.remove();
            if (first == null && second == null) {
                continue;
            }
            if (first == null || second == null || first.val != second.val) {
                return false;
            }

            queue.add(first.left);
            queue.add(second.right);

            queue.add(first.right);
            queue.add(second.left);
        }
        return true;
    }
}
// @lc code=end
